File: collatz.fs
\ Tested in GNU Forth 0.7.3
: collatz ( n -- collatz n ) recursive
  dup .
  dup 1 = if 1
  else
    dup 2 mod 0 = if 2 /mod collatz
    else
      3 * 1 + collatz
    endif
  endif ;

File: collatz.ml
(* Tested with OCaml 4.04.0 *)
let rec collatz n =
  match n with
  | 1 -> 1
  | n -> Printf.printf "%d\n" n ; if n mod 2 = 0 then collatz (n / 2) else collatz (3 * n + 1)
  ;;
